# Machine Learning Training

In this section, I will use the cleaned heart disease dataset to build and compare two machine learning models: Linear Regression and Random Forest.

In [2]:
import pandas as pd

# Load the cleaned dataset
df = pd.read_csv("../data/processed/heart_disease_cleaned.csv")

# Display the first five rows
df.head()

,Age,Gender,Blood Pressure,Cholesterol Level,Exercise Habits,Smoking,Family Heart Disease,Diabetes,BMI,High Blood Pressure,...,Sleep Hours,Sugar Consumption,Triglyceride Level,Fasting Blood Sugar,CRP Level,Homocysteine Level,Heart Disease Status,Mean_Age,Mean_Cholesterol,Cholesterol_Z
0,56.0,Male,153.0,155.0,High,Yes,Yes,No,24.991591,Yes,...,7.633228,Medium,342.0,120.0,12.969246,12.387250,No,49.139387,224.882119,-1.618712
1,69.0,Female,146.0,286.0,High,No,Yes,Yes,25.221799,No,...,8.744034,Medium,133.0,157.0,9.355389,19.298875,No,49.452792,225.977300,1.392213
2,46.0,Male,126.0,216.0,Low,No,No,No,29.855447,No,...,4.440440,Low,393.0,92.0,12.709873,11.230926,No,49.139387,224.882119,-0.216679
3,32.0,Female,122.0,293.0,High,Yes,Yes,No,24.130477,Yes,...,5.249405,High,293.0,94.0,12.509046,5.961958,No,49.452792,225.977300,1.553102
4,60.0,Male,166.0,242.0,Low,Yes,Yes,Yes,20.486289,Yes,...,7.030971,High,263.0,154.0,10.381259,8.153887,No,49.139387,224.882119,0.380910


In [3]:
#Check the cleaned dataset
print("Dataset shape:", df.shape)

print("\nColumns:")
print(df.columns.tolist())

print("\nMissing values:")
print(df.isna().sum())

print("\nData types:")
print(df.dtypes)

Dataset shape: (10000, 24)

Columns:
['Age', 'Gender', 'Blood Pressure', 'Cholesterol Level', 'Exercise Habits', 'Smoking', 'Family Heart Disease', 'Diabetes', 'BMI', 'High Blood Pressure', 'Low HDL Cholesterol', 'High LDL Cholesterol', 'Alcohol Consumption', 'Stress Level', 'Sleep Hours', 'Sugar Consumption', 'Triglyceride Level', 'Fasting Blood Sugar', 'CRP Level', 'Homocysteine Level', 'Heart Disease Status', 'Mean_Age', 'Mean_Cholesterol', 'Cholesterol_Z']

Missing values:
Age                     0
Gender                  0
Blood Pressure          0
Cholesterol Level       0
Exercise Habits         0
Smoking                 0
Family Heart Disease    0
Diabetes                0
BMI                     0
High Blood Pressure     0
Low HDL Cholesterol     0
High LDL Cholesterol    0
Alcohol Consumption     0
Stress Level            0
Sleep Hours             0
Sugar Consumption       0
Triglyceride Level      0
Fasting Blood Sugar     0
CRP Level               0
Homocysteine Level      

In [4]:
# Convert Heart Disease Status into a numerical target

df["Heart_Disease_Target"] = df["Heart Disease Status"].map({
    "No": 0,
    "Yes": 1
})

print(df[["Heart Disease Status", "Heart_Disease_Target"]].head())

  Heart Disease Status  Heart_Disease_Target
0                   No                     0
1                   No                     0
2                   No                     0
3                   No                     0
4                   No                     0


In [5]:
# Select the same features for both models

features = [
    "Age",
    "Blood Pressure",
    "Cholesterol Level",
    "BMI",
    "Sleep Hours"
]

X = df[features]
y = df["Heart_Disease_Target"]

print("Features (X):")
print(X.head())

print("\nTarget (y):")
print(y.head())

Features (X):
    Age  Blood Pressure  Cholesterol Level        BMI  Sleep Hours
0  56.0           153.0              155.0  24.991591     7.633228
1  69.0           146.0              286.0  25.221799     8.744034
2  46.0           126.0              216.0  29.855447     4.440440
3  32.0           122.0              293.0  24.130477     5.249405
4  60.0           166.0              242.0  20.486289     7.030971

Target (y):
0    0
1    0
2    0
3    0
4    0
Name: Heart_Disease_Target, dtype: int64


## Train-Test Split

I split the dataset into 80% training data and 20% testing data. The training data will be used to train the models, while the testing data will be used to evaluate their performance. I use `stratify=y` to maintain the same proportion of heart disease cases in both datasets.

In [7]:
from sklearn.model_selection import train_test_split

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)

print("Training features:", X_train.shape)
print("Testing features:", X_test.shape)
print("Training target:", y_train.shape)
print("Testing target:", y_test.shape)

print("Training target distribution:")
print(y_train.value_counts())

print("\nTesting target distribution:")
print(y_test.value_counts())

Training features: (8000, 5)
Testing features: (2000, 5)
Training target: (8000,)
Testing target: (2000,)
Training target distribution:
Heart_Disease_Target
0    6400
1    1600
Name: count, dtype: int64

Testing target distribution:
Heart_Disease_Target
0    1600
1     400
Name: count, dtype: int64


In [9]:
print("Training features:", X_train.shape)
print("Testing features:", X_test.shape)

print("\nTraining target distribution:")
print(y_train.value_counts())

print("\nTesting target distribution:")
print(y_test.value_counts())

Training features: (8000, 5)
Testing features: (2000, 5)

Training target distribution:
Heart_Disease_Target
0    6400
1    1600
Name: count, dtype: int64

Testing target distribution:
Heart_Disease_Target
0    1600
1     400
Name: count, dtype: int64


## Linear Regression

Linear Regression is used as the first model and baseline for comparison. The model learns the relationship between the five selected health features and the numerical heart disease target (0 = No, 1 = Yes).

In [10]:
from sklearn.linear_model import LinearRegression

# Create the Linear Regression model
linear_model = LinearRegression()

# Train the Linear Regression model
linear_model.fit(X_train, y_train)

# Make predictions on the test data
linear_predictions = linear_model.predict(X_test)

print("First 10 Linear Regression predictions:")
print(linear_predictions[:10])

First 10 Linear Regression predictions:
[0.21182192 0.19687121 0.18622081 0.18510372 0.21369408 0.21066152
 0.19923822 0.20766195 0.2098163  0.19502111]


In [12]:
# Convert predictions into Yes/No  
#logic is Prediction < 0.5  → 0 → No or Prediction ≥ 0.5  → 1 → Yes

linear_classes = (linear_predictions >= 0.5).astype(int)

print("First 10 predicted classes:")
print(linear_classes[:10])

First 10 predicted classes:
[0 0 0 0 0 0 0 0 0 0]


## Evaluate Linear Regression

In [13]:
from sklearn.metrics import accuracy_score, mean_squared_error

linear_accuracy = accuracy_score(
    y_test,
    linear_classes
)

linear_mse = mean_squared_error(
    y_test,
    linear_predictions
)

print("Linear Regression Accuracy:", round(linear_accuracy, 4))
print("Linear Regression MSE:", round(linear_mse, 4))

Linear Regression Accuracy: 0.8
Linear Regression MSE: 0.1605


In [14]:
comparison = pd.DataFrame({
    "Actual": y_test.values,
    "Linear_Prediction": linear_predictions,
    "Predicted_Class": linear_classes
})

comparison.head(10)

,Actual,Linear_Prediction,Predicted_Class
0,0,0.211822,0
1,1,0.196871,0
2,0,0.186221,0
3,0,0.185104,0
4,0,0.213694,0
5,0,0.210662,0
6,0,0.199238,0
7,0,0.207662,0
8,1,0.209816,0
9,1,0.195021,0


## Random Forest

Random Forest is the second model. It uses multiple decision trees to learn patterns between the selected health features and heart disease status. The model uses the same features and train-test split as Linear Regression so that the comparison is fair.

In [17]:
from sklearn.ensemble import RandomForestClassifier

# Create the Random Forest model
random_forest = RandomForestClassifier(
    n_estimators=100,
    random_state=42
)

# Train the Random Forest model
random_forest.fit(X_train, y_train)


rf_predictions = random_forest.predict(X_test)

rf_accuracy = accuracy_score(y_test, rf_predictions)

print("Random Forest Accuracy:", round(rf_accuracy, 4))

Random Forest Accuracy: 0.8005


## Model Comparison

The Linear Regression model achieved an accuracy of 80.00%, while the Random Forest model achieved an accuracy of 80.05%.

Random Forest performed slightly better based on accuracy, but the difference was only 0.05 percentage points. Therefore, the two models performed almost the same on this dataset and with the selected features.

## Confusion Matrix

In [18]:
from sklearn.metrics import confusion_matrix, classification_report

print("Linear Regression Confusion Matrix:")
print(confusion_matrix(y_test, linear_classes))

print("\nLinear Regression Classification Report:")
print(classification_report(y_test, linear_classes))

Linear Regression Confusion Matrix:
[[1600    0]
 [ 400    0]]

Linear Regression Classification Report:
              precision    recall  f1-score   support

           0       0.80      1.00      0.89      1600
           1       0.00      0.00      0.00       400

    accuracy                           0.80      2000
   macro avg       0.40      0.50      0.44      2000
weighted avg       0.64      0.80      0.71      2000



c:\Users\user\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\metrics\_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
c:\Users\user\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\metrics\_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
c:\Users\user\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\metrics\_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(averag

In [19]:
print("Random Forest Confusion Matrix:")
print(confusion_matrix(y_test, rf_predictions))

print("\nRandom Forest Classification Report:")
print(classification_report(y_test, rf_predictions))

Random Forest Confusion Matrix:
[[1597    3]
 [ 396    4]]

Random Forest Classification Report:
              precision    recall  f1-score   support

           0       0.80      1.00      0.89      1600
           1       0.57      0.01      0.02       400

    accuracy                           0.80      2000
   macro avg       0.69      0.50      0.45      2000
weighted avg       0.76      0.80      0.72      2000



Random Forest achieved slightly higher accuracy (80.05%) than Linear Regression (80.00%). However, the difference is very small, so neither model can be considered clearly superior based on accuracy alone. Because the target classes are imbalanced, confusion matrix and classification metrics should also be considered when evaluating the models.